content/blog/debugging-valgrind-gdb.md

```diff
@@ -0,0 +1,190 @@
+---
+title: "Debugging memory errors with valgrind and gdb"
+author: Rich FitzJohn
+date: 2021-06-10
+tags:
+ - dust
+ - R
+ - debugging
+---
+
+We've already written about debugging R packages with valgrind and gdb together [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  This will likely only work on Linux (not windows or macOS).
+
+## The problem
+
+We noticed that as of 0.9.3 of `dust` we had a `std::bad_alloc` error when running models with the `float` type (rather than `double`). After a bit of narrowing it down we managed to reduce the problem to this bit of code:
```

Comment:
Any easy way to explain how it was narrowed down? I think this is something that can be difficult in R

Reply:
Hmm, not much to add - I just took my crashing code and knocked out lines until I got it as simple as I wanted. The original code was from our benchmarking explorations. Or was there something else you meant specifically?

The seed thing is likely not surprising to people using R as it's easy enough to configure the seed there

Reply:
Perhaps just that 'I just took my crashing code and knocked out lines until I got it as simple as I wanted'